File: flutter_app/lib/widgets/draggable_ai_orb.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import '../services/prefs.dart';
import 'ai_orb.dart';

/// A free-floating EZ AI ball. Long-press to pick it up and drag it anywhere;
/// where you drop it is remembered across app restarts. Tap opens EZ AI.
/// Renders as a [Positioned] — place it directly inside a full-screen [Stack].
class DraggableAiOrb extends StatefulWidget {
  const DraggableAiOrb({super.key, required this.onTap});
  final VoidCallback onTap;

  static const double box = 58; // footprint incl. glow
  static const Offset defaultFrac = Offset(0.5, 0.80);

  @override
  State<DraggableAiOrb> createState() => _DraggableAiOrbState();
}

class _DraggableAiOrbState extends State<DraggableAiOrb> {
  Offset? _frac;
  bool _dragging = false;
  bool _loaded = false;

  @override
  void initState() {
    super.initState();
    Prefs.loadOrb().then((f) {
      if (mounted) {
        setState(() {
          _frac = f ?? DraggableAiOrb.defaultFrac;
          _loaded = true;
        });
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    if (!_loaded) return const SizedBox.shrink();

    final media = MediaQuery.of(context);
    final size = media.size;
    const b = DraggableAiOrb.box;
    final frac = _frac!;

    final topSafe = media.padding.top + 6;
    final bottomSafe = media.padding.bottom + 6;
    final left = (frac.dx * size.width - b / 2).clamp(6.0, size.width - b - 6.0);
    final top = (frac.dy * size.height - b / 2).clamp(topSafe, size.height - b - bottomSafe);

    return AnimatedPositioned(
      duration: _dragging ? Duration.zero : const Duration(milliseconds: 260),
      curve: Curves.easeOutCubic,
      left: left,
      top: top,
      width: b,
      height: b,
      child: GestureDetector(
        behavior: HitTestBehavior.opaque,
        onTap: widget.onTap,
        onPanStart: (_) {
          HapticFeedback.lightImpact();
          setState(() => _dragging = true);
        },
        onPanUpdate: (d) {
          setState(() {
            _frac = Offset(
              (d.globalPosition.dx / size.width).clamp(0.0, 1.0),
              (d.globalPosition.dy / size.height).clamp(0.0, 1.0),
            );
          });
        },
        onPanEnd: (_) {
          setState(() => _dragging = false);
          if (_frac != null) Prefs.saveOrb(_frac!);
        },
        onLongPressStart: (_) {
          HapticFeedback.mediumImpact();
          setState(() => _dragging = true);
        },
        onLongPressMoveUpdate: (d) {
          setState(() {
            _frac = Offset(
              (d.globalPosition.dx / size.width).clamp(0.0, 1.0),
              (d.globalPosition.dy / size.height).clamp(0.0, 1.0),
            );
          });
        },
        onLongPressEnd: (_) {
          setState(() => _dragging = false);
          if (_frac != null) Prefs.saveOrb(_frac!);
        },
        child: AnimatedScale(
          duration: const Duration(milliseconds: 200),
          scale: _dragging ? 1.18 : 1.0,
          child: Center(child: AiOrb(size: _dragging ? 44 : 38)),
        ),
      ),
    );
  }
}

File: android_native/app/src/main/java/com/ezbillify/ezworkspace/ui/components/DraggableAiOrb.kt
package com.ezbillify.ezworkspace.ui.components

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import com.ezbillify.ezworkspace.services.Prefs
import kotlinx.coroutines.launch

private val boxSize = 58.dp
private const val DEFAULT_FRAC_X = 0.5f
private const val DEFAULT_FRAC_Y = 0.80f

/**
 * A free-floating EZ AI ball. Long-press-drag to pick it up and move it
 * anywhere; where you drop it is remembered across app restarts. Tap opens
 * EZ AI. Place directly inside a full-screen [Box] (parent must report its
 * size via [containerSize], obtained from [rememberContainerSize]).
 */
@Composable
fun DraggableAiOrb(containerSize: IntSize, onTap: () -> Unit) {
    val context = LocalContext.current
    val prefs = remember { Prefs(context) }
    val scope = rememberCoroutineScope()
    val density = LocalDensity.current

    var fracX by remember { mutableFloatStateOf(DEFAULT_FRAC_X) }
    var fracY by remember { mutableFloatStateOf(DEFAULT_FRAC_Y) }
    var loaded by remember { mutableStateOf(false) }
    var dragging by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        val saved = prefs.loadOrb()
        if (saved != null) {
            fracX = saved.first
            fracY = saved.second
        }
        loaded = true
    }

    if (!loaded || containerSize.width == 0) return

    val boxPx = with(density) { boxSize.toPx() }
    val left = (fracX * containerSize.width - boxPx / 2f).coerceIn(6f, containerSize.width - boxPx - 6f)
    val top = (fracY * containerSize.height - boxPx / 2f).coerceIn(6f, containerSize.height - boxPx - 6f)

    val animatedScale by animateDpAsState(
        targetValue = if (dragging) 56.dp else 52.dp,
        animationSpec = tween(200),
        label = "orbScale",
    )

    Box(
        modifier = Modifier
            .offset(with(density) { left.toDp() }, with(density) { top.toDp() })
            .size(boxSize)
            .pointerInput(containerSize) {
                detectTapGestures(onTap = { onTap() })
            }
            .pointerInput(containerSize) {
                detectDragGestures(
                    onDragStart = { dragging = true },
                    onDragEnd = {
                        dragging = false
                        scope.launch { prefs.saveOrb(fracX, fracY) }
                    },
                    onDragCancel = { dragging = false },
                ) { change, dragAmount ->
                    change.consume()
                    fracX = ((fracX * containerSize.width + dragAmount.x) / containerSize.width).coerceIn(0f, 1f)
                    fracY = ((fracY * containerSize.height + dragAmount.y) / containerSize.height).coerceIn(0f, 1f)
                }
            },
    ) {
        AiOrb(size = animatedScale)
    }
}

/** Reports a Box's pixel size for [DraggableAiOrb]'s positioning math. */
@Composable
fun rememberContainerSize(): Pair<IntSize, Modifier> {
    var size by remember { mutableStateOf(IntSize.Zero) }
    val modifier = Modifier.onSizeChanged { size = it }
    return size to modifier
}
